docs/proposals/Periodic-Retrieval-of-Certificates-and-Keys-from-AKV.md

```diff
@@ -0,0 +1,121 @@
+
+### Certificate revocation
+
+A malicious user compromised specific versions of certificate. Alice promptly reported this incident to the Certificate Authority (CA). The CA added the compromised versions to the Certificate Revocation List (CRL) and responded to Online Certificate Status Protocol (OCSP) queries with the revoked status. As a result, images signed with the revoked versions fail signature verification. Alice rotated the certificate, obtaining a new version issued by the CA while maintaining the same certificate name and identity information. The build pipeline was triggered to build and sign all the images with the new version of the certificate. Furthermore, as Alice configures Ratify to periodically retrieve all available versions of the certificate from AKV, it ensures successful verification of images signed with the latest version without requiring manual configuration adjustments. By following these actions, Alice maintains security while seamlessly transitioning to the new certificate version.
+
+### Key rotation
+
+Bob, an application engineer at Wabbit Network LLC, focuses on securing containerized applications. He sets up the build pipeline to sign container images using Cosign and key pairs in AKV. To ensure that only trusted images are deployed in Kubernetes clusters, he deploys Gatekeeper and Ratify. Ratify is configured to validate Cosign signatures using keys from AKV. As a security best practice, Bob configured automated key rotation for the key in AKV, resulting a new version of the key created before key expiry. He ensures that the pipeline always uses the latest key version for signing container images. Additionally, Bob configures Ratify to periodically retrieve all available versions of the key from AKV, allowing successful verification of images signed by different key versions without any manual configuration adjustments. Bob knows that images are not allowed to be signed with Cosign using keys that have expired in AKV, but verification using expired key is allowed. He assumes that the old versions of the keys are retrieved by Ratify, because there are still some images around that were signed with old versions.
```

Comment:
great point. The rotation must be disabled by default in v1.3 as this would be a breaking change.  Secret store csi driver has a global param [--rotation-poll-interval](https://secrets-store-csi-driver.sigs.k8s.io/topics/secret-auto-rotation) that defaults to 2m.  Do you think this is where we could start ? @akashsinghal , @yizha1 , @duffney .  Note for secret store csi driver, the rotation feature is still in alpha after its been available for >2yrs. Not sure if that is a indicator on usage.

Reply:
> `Bob configured automated key rotation for the key in AKV` Does this mean automatic fetching will be opt in? What granularity will be enforced at? Per KMP? Global Ratify level? Per key/cert?

This is the security best practice, to rotate key or certificate regularly. If users do not rotate the key, then there is no need to periodic retrieval of new versions. But there is still a need to regularly reconcile, because a key can be disabled for some reasons.

From users' point of view, I want to configure Ratify KMP once, and then Ratify can keep all the versions up-to-date. Otherwise, I need to specify a version manually. This is exactly the problem we want to solve. I would say per KMP, and specifically, for AKV as the KMP.

Reply:
@binbin-li Rotation will improve the security posture. Assuming you use one and only key to sign all the images from time to time, and if the one and only key is compromised, then all the images are compromised. If you rotated the key regularly, and a specific version is compromised, then only images signed with the specific version are compromised. Other images are all good. After rotation, users should set up the build system to use the latest version in a proper time to sign images, however, the previous version will not expire immediately, so there is also a period that some images are signed with a previous version, some are signed with the latest version. Ultimately all the new images are signed with latest version, the images signed with old keys will be end of life gradually, as old images can be vulnerable. 

For Notary Project signature, with timestamping support, users can verify images that are signed before the certificate expired.  For cosign public key verification, as KMS ensures that expired key cannot be used for signing, so it is OK to use expired key for verification, users can disable the key in the KMS if they do not want it to be used any more.  Please be noted, it is recommended to use a KMS as `KMP`, for inline provider, as there is no way to check the key status from a KMS, so it is not recommended for production.

Reply:
Resolved as discussed during community meeting on Jun 12 2024.